### ISSUE TITLE
LightGBMTunerCV doing wrong early stopping and gives wrong model at end

### ISSUE THREAD

**Issue by pseudotensor** [optuna/optuna#3631]
### Expected behavior

Using https://github.com/optuna/optuna-examples/blob/main/lightgbm/lightgbm_tuner_cv.py
and following this:
https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.lightgbm.LightGBMTunerCV.html

Three (related?) problems:

1) A single training period does not do reasonable early stopping.  Instead, of 100 early stopping rounds, training continues after the best score for another 500 etc. iterations consuming all boosting rounds.  Optuna seems to see the final score as the best one for that trial, or just bad reporting.  But this is definitely a waste.  I understand optuna pruner can terminate things even earlier than early stopping, but why are things so delayed beyond the early stopping rounds period beyond the best score?

i.e. one will see:
```
[899]   cv_agg's multi_logloss: 8.88302 + 12.3886
[900]   cv_agg's multi_logloss: 8.88302 + 12.3886
[900]   cv_agg's multi_logloss: 8.88302 + 12.3886
[901]   cv_agg's multi_logloss: 8.88304 + 12.3886
[902]   cv_agg's multi_logloss: 8.88304 + 12.3886
[903]   cv_agg's multi_logloss: 8.88304 + 12.3886
[904]   cv_agg's multi_logloss: 8.88303 + 12.3886
[905]   cv_agg's multi_logloss: 8.88304 + 12.3886
[906]   cv_agg's multi_logloss: 8.88304 + 12.3886
/home/jon/minicondadai_py38/lib/python3.8/site-packages/lightgbm/basic.py:2065: UserWarning: Using categorical_feature in Dataset.
  _log_warning('Using categorical_feature in Dataset.')
/home/jon/minicondadai_py38/lib/python3.8/site-packages/lightgbm/engine.py:620: UserWarning: 'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. Pass 'log_evaluation()' callback via 'callbacks' argument instead.
  _log_warning("'verbose_eval' argument is deprecated and will be removed in a future release of LightGBM. "
[I 2022-06-04 14:53:16,558] The booster of trial#24 was saved as ./models2/24.pkl.
^Mnum_leaves, val_score: 0.122860:  85%|########5 | 17/20 [36:17<06:11, 123.89s/it]^Mnum_leaves, val_score: 0.122860:  90%|######### | 18/20 [36:17<04:15, 127.56s/it][I 2022-06-04 14:53:16,626] Trial 24 finished with value: 8.882802609368346
```

i.e. `Trial 24 finished with value: 8.882802609368346` but the last trial going to 906 iterations actually had a best score at a much earlier iteration
```
[80]    cv_agg's multi_logloss: 0.237891 + 0.00906186
```

Why is it saying trial finished with very last value instead of best in the trial?  Not sure this is bug, but the excessive iterations seems to be a bug.

2) The 
```
Did not meet early stopping. Best iteration is:
[998]   cv_agg's multi_logloss: 0.180995 + 0.0117511
```

type report always refers to (seemingly) the global best, not the last training round.  This is confusing, or may be a bug of some kind.  Or it may be intentional, but it may be related to the other problems here.

3) The final manifestation of the above or yet another bug is that at the end the the report might look  like:

```
[834]   cv_agg's multi_logloss: 8.7975 + 6.13048
[835]   cv_agg's multi_logloss: 8.77788 + 6.11675
[836]   cv_agg's multi_logloss: 8.7437 + 6.09312
[837]   cv_agg's multi_logloss: 8.72923 + 6.08319
[838]   cv_agg's multi_logloss: 8.77614 + 6.11553
[839]   cv_agg's multi_logloss: 8.83346 + 6.1558
[840]   cv_agg's multi_logloss: 8.82765 + 6.15171
[841]   cv_agg's multi_logloss: 8.79684 + 6.12999
[842]   cv_agg's multi_logloss: 8.79434 + 6.12822
[843]   cv_agg's multi_logloss: 8.75756 + 6.10263
[844]   cv_agg's multi_logloss: 8.75599 + 6.10153
Early stopping, best iteration is:
[744]   cv_agg's multi_logloss: 0.122506 + 0.00420387
Best score: 0.12250622939778731
Best params: {'verbosity': -1, 'boosting_type': 'gbdt', 'n_jobs': 10, 'device_type': 'cpu', 'objective': 'multiclass', 'metric': 'multi_logloss', 'num_classes': 7, 'feature_pre_filter': False, 'lambda_l1': 0.0, 'lambda_l2': 0.0, 'num_leaves': 
  Params: 
    verbosity: -1
    boosting_type: gbdt
    n_jobs: 10
    device_type: cpu
    objective: multiclass
    metric: multi_logloss
    num_classes: 7
    feature_pre_filter: False
    lambda_l1: 0.0
    lambda_l2: 0.0
    num_leaves: 149
    feature_fraction: 0.7
    bagging_fraction: 1.0
    bagging_freq: 0
    min_child_samples: 20
best_iteration: 744
```

However, if one takes those best parameters and builds a new model for those best iterations, the score is very poor of 8.7908 logloss, which is quite close to the very last round.  Something seems messed up with the best parameters.

It's like every training round is replacing the best even if not, which may be related to issue 1 and 2.

So overall both major accuracy (totally wrong or bad model) and performance (wasted time) issues.

### Environment

```
- Optuna version:2.10.0
- Python version:3.8.12
- OS:Linux-5.4.0-74-generic-x86_64-with-glibc2.10
```

### Error messages, stack traces, or logs

```shell
No errors, logs could be included but there is no way to attach a file here.
```

### Steps to reproduce

```python
import os

import numpy as np

import sklearn
from sklearn.base import BaseEstimator, ClassifierMixin
from sklearn.preprocessing import LabelEncoder
from sklearn.utils.multiclass import unique_labels


def get_data(task_id):
    import openml
    cache_dir = "openml_benchmark"
    openml.config.cache_directory = os.path.expanduser(cache_dir)

    task = openml.tasks.get_task(task_id)  # download the OpenML task
    labels = task.class_labels

    train_indices, test_indices = task.get_train_test_split_indices()
    X, y = task.get_X_and_y(dataset_format="dataframe")
    X_train = X.iloc[train_indices]
    y_train = y.iloc[train_indices]
    X_test = X.iloc[test_indices]
    y_test = y.iloc[test_indices]

    return X_train, y_train, X_test, y_test, labels


def go():
    X_train, y_train, X_test, y_test, labels = get_data(75193)

    model = OptunaLightGBMClassifier()
    model.fit(X_train, y_train)

    preds = model.predict_proba(X_test)
    score = sklearn.metrics.log_loss(y_test, preds, labels=labels)
    print(score)


class OptunaLightGBMClassifier(BaseEstimator, ClassifierMixin):
    def fit(self, X, y):
        X_numeric = X.select_dtypes(include=np.number)
        num_cols = sorted(X_numeric.columns)
        cat_cols = sorted(set(X.columns).difference(num_cols))
        from sklearn.preprocessing import OrdinalEncoder
        self._cat_encs = {}
        for col in cat_cols:
            self._cat_encs[col] = OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1,
                                                 encoded_missing_value=-1)
            X.loc[:, col] = self._cat_encs[col].fit_transform(X[[col]])

        self.classes_ = unique_labels(y)
        self._le = LabelEncoder()
        y = self._le.fit_transform(y)
        self.n_features_in_ = X.shape[1]

        import optuna.integration.lightgbm as lgboptuna
        from lightgbm import early_stopping
        from lightgbm import log_evaluation
        from sklearn.model_selection import KFold

        data, target = X, y

        dtrain = lgboptuna.Dataset(data, label=target, categorical_feature=cat_cols)

        n_jobs = 10
        params = {
            "verbosity": -1,
            "boosting_type": "gbdt",
            "n_jobs": n_jobs,
            'device_type': 'cpu',
            'deterministic': True,
            "objective": "multiclass",
            "metric": "multiclass",
            'num_classes': len(self.classes_)
        }

        seed = 2
        model_dir = "./models%s/" % seed
        os.makedirs(model_dir)
        tuner = lgboptuna.LightGBMTunerCV(
            params,
            dtrain,
            folds=KFold(n_splits=3),
            seed=seed,
            num_boost_round=1000,
            callbacks=[early_stopping(100), log_evaluation(100)],
            time_budget=60 * 60,
            return_cvbooster=seed != 0,
            optuna_seed=seed,
            model_dir=model_dir,
        )

        tuner.run()

        print("Best score:", tuner.best_score)
        best_params = tuner.best_params
        print("Best params:", best_params)
        print("  Params: ")
        for key, value in best_params.items():
            print("    {}: {}".format(key, value))

        # fit final sklearn model with no early stopping and no CV so sees all data
        import lightgbm as lgborig
        best_iteration = tuner.get_best_booster().best_iteration
        print("best_iteration: %s" % best_iteration)
        best_params.update(n_estimators=best_iteration + 1, n_jobs=n_jobs)
        best_params.update(seed=seed, random_state=seed)

        self._model = lgborig.LGBMClassifier(**best_params)
        self._model.fit(X, y, categorical_feature=cat_cols)

    def predict_proba(self, X):
        for col, enc in self._cat_encs.items():
            X.loc[:, col] = self._cat_encs[col].transform(X[[col]])

        preds = self._model.predict_proba(X)
        return preds

    def get_params(self, deep=True):
        return {}

    def set_params(self, **params):
        for k, v in params.items():
            setattr(self, k, v)
        return self


if __name__ == "__main__":
    go()

```

**Comment by pseudotensor** [comment]
Logs showing the issues 1 and 2:

[tunerrepro1.log](https://github.com/optuna/optuna/files/8838877/tunerrepro1.log)


Logs for issue 3, didn't yet repro with just the code given.

[badmodel.txt](https://github.com/optuna/optuna/files/8838878/badmodel.txt)

**Comment by pseudotensor** [comment]
Here's a case of the issue 3 with actual logs from the script above, using seed=10.

[seed10.log](https://github.com/optuna/optuna/files/8839317/seed10.log)


Minor repro code changes:

[tunerrepro1.py.zip](https://github.com/optuna/optuna/files/8839318/tunerrepro1.py.zip)

**Comment by pseudotensor** [comment]
I think this is a problem with LightGBM, not Optuna, in translation from native API to sklearn API.  If I fit with the native API there is no such issue.

I know in the past there have been inconsistencies: https://github.com/microsoft/LightGBM/issues/1033 but that was in 2017.  But I guess there are others still (i'm using latest optuna and latest sklearn from pypi).

It's still important to Optuna because it means the tuned "parameters" are not sufficient, there are hidden values not stored in the booster that the optuna-lightgbm package should manage better.

**Comment by nzw0301** [comment]
I'm wondering if the issues (at least a part of issues) could be caused by the same reason as https://github.com/optuna/optuna/issues/3145.

**Comment by pseudotensor** [comment]
Ya sounds plausible.  It's definitely not doing early stopping sometimes.

The "bad model" issue is not due to optuna, but lightgbm: https://github.com/microsoft/LightGBM/issues/5268 and some kind of seed instability.

So for Optuna, main question is why aren't the callbacks respected always?  I see sometimes early stopping, and other times not.

Is this a possible bug in LightGBM only with the callbacks?  Or is it some interaction between the two?

**Comment by pseudotensor** [comment]
fixed if use latest lightgbm/optuna and some fixes mentioned in linked issue above.
